doc/man-source/irsimreceive.1
.TH irsimreceive "1" "Last change: Oct 2015" "irsimreceive @version@" "User Commands"
.SH NAME
.P
\fBirsimreceive\fR - Decode IR durations data.
.SH SYNOPSIS
.P
\fBirsimreceive\fR [\fIoptions\fR]  <\fIconfigfile\fR>  <\fIdatafile\fR>

.SH DESCRIPTION
.P
This is a simple test tool which decodes duration data in \fIdatafile\fR
using an lircd.conf type \fIconfigfile\fR. The outputted data is
in the same format as on the lircd output socket.

.SH OPTIONS
.TP 4
.B -U, --plugindir <\fIdir\fR>
Load drivers from <\fIdir\fR>.

.TP 4
.B -v , --version
Print version and exit.

.TP 4
.B -h , --help
Print help message.

.SH FILES
.TP 4
.B ~/.cache/irsimreceive.log
Debug log. Setting the XDG_CACHE_HOME environment variable relocates this
file to $XDG_CACHE_HOME/irsimreceive.log. The amount of logging respects the
LIRC_LOGLEVEL environment variable, defaulting to the \fidebug\fR value
in the \fi[lircd]\fR  section of lirc_options.conf.


.SH "SEE ALSO"
.P
The documentation for \fBlirc\fR
is maintained as html pages. They are located under html/ in the
documentation directory.


NEWS
0.9.4-devel
  * The caraca driver has been retired (still available in attic/).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * Enhanced lircd-setup tool (#123)
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers.
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.
    - Data is collected from drivers, remotes and static files.
    - Better ways for drivers to enumerate devices.
    - There is an unstable python database module used by
      all tools.
  * devinput.lircd.conf is generated at build time to match host kernel.
  * A new script lircd-setup runs as a one-shot service before lircd,
    performing setup commands as root (#121).

  * CLI options for mode2, xmode2 and irrecord has been streamlined.
    All tools falls back to lircd defaults where it makes sense.
  * New logging API with hooks for enhanced messages and filtering
    (also #129).
  * The out-of-tree build interface is enhanced with hooks for external
    drivers to rebuild documentation and add a config file.
  * irrecord has benn taught to handle remotes without gap (#98).
  * Add an "auto" device to devinput driver for single-device
    usecase (#134).

  * Fix buggy handling of not decoded buttons in irrecord (#146).
  * Fix --plugindir not being respected (#144).
  * Fix crash for non-existing device (#145).
  * Fix gid and supplementary groups in CLI apps dropping root.
  * Fix a crash in irrecord when invoked without options.
  * Don't require write access if not needed in default.c (#133).
  * Clean up docs, remove circular links (#136). Documentation is
    now only in lirc.org/html.

0.9.3  4/09/15
  * Digest of 0.9.3pre1..0.9.3pre3

  * New build dependency: xsltproc (sometimes coming with libxml).
  * The irexec semantics have been changed so that the invoked
    programs run disconnected from the main irexec process.
    See irexec(1) (#92).
  * irrecord has got some love:
     + It handles multiple definitions of same symbol OK (#78)
     + Warns about single-key config files (#77)
     + Checks for ambient light disturbances (#76)
     + New option allows updating of existing template file.
     + Template handling bug fixed
     + -l --list-namespace option parsing bug fixed (#101).
     + Minor fixes to irrecord-ui including -k option parsing fixed.
  * mode2 has been updated
     +  handle also LIRCCODE devices (#94),
     +  Drop root privileges (#91)
     +  Add  new scope-like display
     +  Handle the--driver-options switch
  * xmode2 has been updated to handle udp driver, the new
    --driver-options switch + some general overhaul.
  * irrecord has been split into a library and a cli interface
    (#36). The library interface is unstable.

  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).
  * lircmd: Several option parsing bugs fixed, one of which
    blocked the systemd service (#104, RH bz 1234155).
  * lirc-lsremotes now displays logs after parsing files, making
    more parsing warnings visible.
  * lirc-config-tool is fixed to comply with guidelines (#105).
  * Bug in irexec when using modes fixed (#102).
  * Create log directory if required (#124).
  * lirc-setup: Fixed name clash on recent python versions.

  * irtoy: New support for debug LEDs. New documentation
    file, support in lirc-setup and driver table doc added.
  * The udp driver is patched to allow adjustable clockrates,
    see new udp_driver.html driver doc.
  * The mplay driver has been updated with patches which has
    been available "out there".

  * The library routines for parsing lircd.conf now warns for
    duplicate keys and values (#107).
  * Enhanced pkg-config support for drivers compiled out-of-tree
    (see driver-api in the manual).
  * Hardcoded CFLAGS in Makefiles fixed (#124).
  * The compiler has been switched to c++, and filenames are
    updated to *.cpp (#79)
  * Build: now uses parallel make by default.
  * Overall code formatting is improved with more consistent
    layout and updated git hook. See CONTRIBUTE.md

  * Unlinked and hard to find driver docs and manpages have
    been linked.
  * The documentation is heavily revised. Old obsolete text is
    removed, the manual is reorganized, some info is moved to
    manpages and a lot of minor updates and bugfixes.

  * The lircd.simsend and lircd.simreceive tools have been removed,
    superseeded by irsimsend and irsimreceive.
  * A new kernel driver irpipe has been added. This allows feeding data
    from userspace to lircd, which can use the /dev/irpipeX device
    as a clean substitute for /dev/lircX. A companion CLI tool
    irpipe(1) is also added.
  * Options parsing (notably -U) and just listing first remote
    bugs in irsimsend have been fixed.


0.9.3pre3 16/08/15
  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).
  * Create log directory if required (#124).
  * The remotes/ directory is not used any more, removed.
  * The lircd.simsend and lircd.simreceive tools have been removed,
    superseeded by irsimsend and irsimreceive.
  * Minor fixes to lirc-setup (bbd4ccf, e384375).
  * Minor fixes to irrecord-ui including -k option parsing fixed.
  * Hardcoded CFLAGS in Makefiles fixed (#124).
  * Enhanced pkg-config support for drivers compiled out-of-tree
    (see driver-api in the manual).


tools/irsimreceive.cpp
/****************************************************************************
** irsimreceive.c **********************************************************
****************************************************************************
*
* irsimreceive.c -Receive data from file and decode.
*
*/

#include <config.h>

#include <stdlib.h>
#include <getopt.h>

#include "lirc_private.h"
#include "lirc_client.h"

static const logchannel_t logchannel = LOG_APP;

static void add_defaults(void)
{
	static char plugindir[128];
        const static char* defaults[] = {
		"lircd:plugindir",	plugindir,
		(const char*)NULL,	(const char*)NULL
	};
	const char* s = getenv("LIRC_PLUGIN_PATH");

	strncpy(plugindir, s != NULL ? s : PLUGINDIR, sizeof(plugindir) - 1);
	options_add_defaults(defaults);
}


static const char* const USAGE =
	"Usage: irsimreceive [options]  <configfile>  <datafile>\n\n"
	"<configfile> is a lircd.conf type configuration.\n"
	"<datafile> is a list of pulse/space durations.\n\n"
	"Options:\n"
	"    -U, --plugindir <path>:     Load drivers from <path>.\n"
	"    -v, --version               Print version.\n"
	"    -h, --help                  Print this message.\n";

static struct option options[] = {
	{ "help",	no_argument,	   NULL, 'h' },
	{ "version",	no_argument,	   NULL, 'v' },
	{ "pluginpath", required_argument, NULL, 'U' },
	{ 0,		0,		   0,	 0   }
};


static void parse_options(int argc, char** const argv)
{
	long c;

	add_defaults();

	while ((c = getopt_long(argc, argv, "hvU:", options, NULL))
	       != EOF) {
		switch (c) {
		case 'h':
			fputs(USAGE, stdout);
			exit(EXIT_SUCCESS);
		case 'v':
			printf("%s\n", "irw " VERSION);
			exit(EXIT_SUCCESS);
		case 'U':
			options_set_opt("lircd:plugindir", optarg);
			break;
		case '?':
			fprintf(stderr, "unrecognized option: -%c\n", optopt);
			fputs("Try `irsimsend -h' for more information.\n",
			      stderr);
			exit(EXIT_FAILURE);
		}
	}
	if (argc != optind + 2) {
		fputs(USAGE, stderr);
		exit(EXIT_FAILURE);
	}
}


static void setup(const char* path)
{
	struct option_t option;
	int r;

	if (access(path, R_OK) != 0) {
		fprintf(stderr, "Cannot open %s for read\n", path);
		exit(EXIT_FAILURE);
	}
	if (hw_choose_driver("file") == -1) {
		fputs("Cannot load file driver (bad plugin path?)\n",
		      stderr);
		exit(EXIT_FAILURE);
	}
	r = curr_driver->open_func("dummy.out");
	if (r == 0) {
		fputs("Cannot open driver\n", stderr);
		exit(EXIT_FAILURE);
	}
	r = curr_driver->init_func();
	if (r == 0) {
		fputs("Cannot init driver\n", stderr);
		exit(EXIT_FAILURE);
	}
	strcpy(option.key, "set-infile");
	strncpy(option.value, path, sizeof(option.value));
	r = curr_driver->drvctl_func(DRVCTL_SET_OPTION, (void*)&option);
	if (r != 0) {
		fputs("Cannot set driver infile.\n", stderr);
		exit(EXIT_FAILURE);
	}
}


struct ir_remote* read_lircd_conf(const char* configfile)
{
	FILE* f;

	struct ir_remote* remotes;
	const char* filename = configfile;

	filename = configfile == NULL ? LIRCDCFGFILE : configfile;
	f = fopen(filename, "r");
	if (f == NULL) {
		log_perror_err("could not open config file '%s'", filename);
		exit(EXIT_FAILURE);
	}
	remotes = read_config(f, configfile);
	fclose(f);
	if (remotes == (void*)-1) {
		log_error("reading of config file failed");
		exit(EXIT_FAILURE);
	} else {
		log_debug("config file read");
		if (remotes == NULL) {
			log_error("config file contains no valid remote control definition");
			exit(EXIT_FAILURE);
		}
	}
	return remotes;
}


void printcode(char* s)
{
	int len;

	if (s == NULL) {
		puts("None");
	} else {
		len = strlen(s);
		if (strlen(s) > 2 && s[len - 1] == '\n')
			s[len - 1] = '\0';
		printf("%s\n", s);
	}
}


int simreceive(struct ir_remote* remotes)
{
	char* code = NULL;
	int at_eof;

	do {
		code = curr_driver->rec_func(remotes);
		at_eof = code != NULL && strstr(code, "__EOF") != NULL;
		if (code != NULL && !at_eof) {
			printcode(code);
			fflush(stdout);
		}
	} while (!at_eof);
	return 0;
}


int main(int argc, char* argv[])
{
	struct ir_remote* remotes;
	char path[128];
	const loglevel_t level = options_get_app_loglevel("irsimreceive");

	lirc_log_get_clientlog("irsimreceive", path, sizeof(path));
	lirc_log_set_file(path);
	lirc_log_open("irsimreceive", 1, level);

	options_load(argc, argv, NULL, parse_options);
	setup(argv[optind + 1]);
	remotes = read_lircd_conf(argv[optind]);
	return simreceive(remotes);
}
